op-service/testlog/capturing.go

```diff
@@ -8,16 +8,67 @@ import (
 	"github.com/ethereum/go-ethereum/log"
 )
 
+// CapturedAttributes forms a chain of inherited attributes, to traverse on captured log records.
+type CapturedAttributes struct {
+	Parent     *CapturedAttributes
+	Attributes []slog.Attr
+}
+
+// Attrs calls f on each Attr in the [CapturedAttributes].
+// Iteration stops if f returns false.
+func (r *CapturedAttributes) Attrs(f func(slog.Attr) bool) {
+	searching := true
+	if r.Parent != nil {
+		r.Parent.Attrs(func(a slog.Attr) bool {
+			searching = f(a)
+			return searching
+		})
+	}
+	if !searching { // if the parent attributes context traversal found it already, then don't traverse the remainder
+		return
+	}
+	for _, a := range r.Attributes {
+		if !f(a) {
+			return
+		}
+	}
+}
+
+// CapturedRecord is a wrapped around a regular log-record,
+// to preserve the inherited attributes context, without mutating the record or reordering attributes.
+type CapturedRecord struct {
+	Parent *CapturedAttributes
+	*slog.Record
+}
+
+// Attrs calls f on each Attr in the [CapturedRecord].
+// Iteration stops if f returns false.
+func (r *CapturedRecord) Attrs(f func(slog.Attr) bool) {
+	searching := true
+	if r.Parent != nil {
+		r.Parent.Attrs(func(a slog.Attr) bool {
+			searching = f(a)
+			return searching
+		})
+	}
+	if !searching { // if the parent attributes context traversal found it already, then don't traverse the remainder
+		return
+	}
+	r.Record.Attrs(f)
```

Comment:
Shouldn't this be the other way around - that the child attributes override the parent's?

Reply:
Good point, I'll fix that